File: android/app/src/main/kotlin/cn/jesse/guideflutter/platform/channel/BasicChannelManager.kt
package cn.jesse.guideflutter.platform.channel

import android.app.Application
import android.text.TextUtils
import android.widget.Toast
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.StandardMessageCodec


/**
 * basic channel 管理器
 *
 * @author jesse
 */
class BasicChannelManager constructor(private var application: Application, private var flutterEngine: FlutterEngine) {
    companion object {
        const val CHANNEL_TOAST = "channel_toast"
        const val CHANNEL_APP_INFO = "channel_app_info"
    }

    /**
     * 统一注册基础功能
     */
    fun registerCommonChannels() {
        registerToast()
        registerAppInfo()
    }

    /**
     * 注册并监听toast
     */
    private fun registerToast() {
        val basicMessageChannel = BasicMessageChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_TOAST, StandardMessageCodec.INSTANCE)
        basicMessageChannel.setMessageHandler { o, _ ->
            val map = o as Map<*, *>;
            if (!TextUtils.isEmpty(map["message"].toString())) {
                val longDuration = map["longDuration"] as Boolean
                Toast.makeText(application, map["message"].toString(), if (longDuration) Toast.LENGTH_LONG else Toast.LENGTH_SHORT).show()
            }

        }
    }

    /**
     * 注册监听回调app信息
     */
    private fun registerAppInfo() {
        val basicMessageChannel = BasicMessageChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_APP_INFO, StandardMessageCodec.INSTANCE)
        basicMessageChannel.setMessageHandler { _, reply ->
            val result = mapOf(
                    "appVersionName" to "v1.0.0",
                    "appVersionCode" to "1",
                    "flutterVersionName" to "v0.5.0",
                    "flutterVersionCode" to "12"
            )
            reply.reply(result)
        }
    }
}

File: lib/platform/channel/basic_channel_manager.dart
import 'package:flutter/services.dart';

/// basic message channel 处理器
class BasicChannelManager {
  static const String CHANNEL_TOAST = 'channel_toast';
  static const String CHANNEL_APP_INFO = 'channel_app_info';

  static const String KEY_TAOST_MESSAGE = 'message';
  static const String KEY_TAOST_DURATION = 'longDuration';

  /// toast信息
  /// toast信息, {是否为长toast}
  static void toast(String message, {bool longDuration = false}) {
    BasicMessageChannel(CHANNEL_TOAST, StandardMessageCodec())
        .send({KEY_TAOST_MESSAGE: message, KEY_TAOST_DURATION: longDuration});
  }

  static const String KEY_APP_INFO_APP_VERSION_NAME = 'appVersionName';
  static const String KEY_APP_INFO_APP_VERSION_CODE = 'appVersionCode';
  static const String KEY_APP_INFO_FLUTTER_VERSION_NAME = 'flutterVersionName';
  static const String KEY_APP_INFO_FLUTTER_VERSION_CODE = 'flutterVersionCode';

  /// 获取原生app信息
  static Future getPlatformAppInfo() {
    return BasicMessageChannel(CHANNEL_APP_INFO, StandardMessageCodec())
        .send(Map());
  }
}
